[WIP]vim-patch:8.1.0953

janlazo opened the pull request:
There's an extra `stdint.h` include in `digraph.c` but modifying that file affects the Vim 5 digraph comments.

justinmk commented on src/nvim/pos.h:
```diff
@@ -1,6 +1,9 @@
 #define NVIM_POS_H
 
+// for INT_MAX, LONG_MAX et al.
+#include <limits.h>
```
unless this is required for things directly in pos.h, in general we don't want to have "mega-headers" that bring in other headers (`os.h` is an exception).

janlazo replied:
It's used directly below it. Do you prefer moving the `MAXCOL`, `MAXLNUM` to another header to avoid an `#include`?

justinmk replied:
No, but maybe worth mentioning in the commit message.

janlazo replied:
`LONG_MAX` is in the patch description. Do you mean to mention that a new header is included in some file?

justinmk replied:
No, I get it now.

justinmk commented on src/nvim/pos.h:
```diff
@@ -11,9 +14,9 @@ typedef int colnr_T;
+#define MAXLNUM LONG_MAX
 /// Maximal column number, 31 bits
-enum { MAXCOL = 0x7fffffff };
+#define MAXCOL INT_MAX
```
what's the reason for this? ZyX made it an enum intentionally.

janlazo replied:
Consistency with MAXLNUM. `enum` is restricted to `int` values.

janlazo commented on src/nvim/pos.h:
```diff
@@ -11,9 +14,9 @@ typedef int colnr_T;
 
 /// Maximal (invalid) line number
-enum { MAXLNUM = 0x7fffffff };
+#define MAXLNUM LONG_MAX
```
This breaks buffer updates.

Tried changing this line to `const long MAXLNUM = LONG_MAX;` but it fails the static assertions where `SIZE_MAX >= MAXLNUM`.

justinmk replied:
cc @phodge @KillTheMule

KillTheMule replied:
So I started looking at this, but I'm confused on an early level. First, the assertions are justified because of the following casts, good we have them :)

But then, if `SIZE_MAX < MAXLNUM`, that means we're potentially having more lines than fit in a `size_t`. Is that a real usecase, can that happen? I'm on shaky ground here, but I always assumed `SIZE_MAX` is really the largest addressable thing. The vim issue doesn't really discuss things...

I'd have figured `#define MAXLUM SIZE_MAX` would be the way to go then. Since line numbers are `long`, things get complicated, but could we just take the smaller value of `LONG_MAX` and `SIZE_MAX`?

We've always left the question open on chunking updates, and I'm a tad unclear on how it could be done.

justinmk replied:
> But then, if `SIZE_MAX < MAXLNUM`, that means we're potentially having more lines than fit in a `size_t`. Is that a real usecase, can that happen?

No, it can't happen.

janlazo committed:
vim-patch:8.1.0953: a very long file is truncated at 2^31 lines

Problem:    A very long file is truncated at 2^31 lines.
Solution:   Use LONG_MAX for MAXLNUM. (Dominique Pelle, closes vim/vim#4011)
https://github.com/vim/vim/commit/c9629251a634d4f5988c8162ba8249026d1af687

janlazo commented:
I give up :disappointed:
